### ISSUE TITLE
Whitelist - Firefox Sync

### ISSUE THREAD

**Issue by rahulgi1** [dillbyrne/random-agent-spoofer#92]
can you whitelist firefox sync 
getting wrong user/pass while using
(only when using github version, no problem with firefox version) 

I have tried all possible whitelist
{"whitelist":[{"url":"accounts.firefox.com","options":["via","date","canvas","ifnonematch","referer","xff","windowName"]}]}

also an option to whitelist tab will be great, rather than a specific site

**Comment by letorbi** [comment]
Once #142 has been merge into the main repo you can use the following whitelist lines to allow Firefox sync:

    [{"url":"firefox.com","options":["authorization"]},{"url":"mozilla.com","options":["authorization"]}]

This think the domain names could be more specific, but I haven't had the time to check it any further.

**Comment by dillbyrne** [comment]
The accounts.firefox.com site also requires  `Disable local dom storage` to be unchecked .

**Comment by dillbyrne** [comment]
@letorbi have you tested the whitelist rule above with firefox sync ?

**Comment by letorbi** [comment]
Yes I've tried. It seems to work for me - even without `local.storage` disabled. However, I just checked whether Firefox shows an error message when trying to connect to the sync server. I didn't test if the data sync actually works.

**Comment by dillbyrne** [comment]
@letorbi if you have a firefox sync account would you mind testing if it syncs the data ? if not I will make an account to test it. 

The local dom storage was needed for the signup page to display the form.

**Comment by letorbi** [comment]
I'll check whether it works or not and think I may have some results tomorrow evening...
On Feb 9, 2015 6:31 PM, dillbyrne <notifications@github.com> wrote:@letorbi if you have a firefox sync account would you mind testing if it syncs the data ? if not I will make an account to test it. 

The local dom storage was needed for the signup page to display the form.

**Comment by letorbi** [comment]
Syncing from the Desktop (RAS enabled) to my mobile (no RAS) works, but vice versa doesn't seem to work. I'll investigate further...

**Comment by dillbyrne** [comment]
@letorbi thank you. I'll leave this issue open until we know more

**Comment by letorbi** [comment]
Hej guys,

I'm back again :) I just did some heavy testing and found out some strange things:

  1. Sync works with all features enabled except for *Disable Authorization header* and whitelist
  2. Sync works with 1.) and the default whitelist enabled
  3. Sync works with 2.) and `{"url":"firefox.com","options":["authorization"]},` added to the whitelist
  4. Sync doesn't work with 3.) and `{"url":"mozilla.com","options":["authorization"]},` added to the whitelist

The strange thing is that Sync does login without problems, but the import of bookmarks from my mobile doesn't seem to work. Apart from that it's even more strange that *Disable Authorization header* was disabled in any test-case, so adding mozilla.com to the whitelist should make no change at all.

All in all it seems that the whiltelist line `{"url":"mozilla.com","options":["authorization"]}` breaks Firefox Sync. I'm going to make some more tests to proof this...

**Comment by letorbi** [comment]
Ok, here's my second series of tests:

  1. Sync works with __all features deactivated*__ - the whitelist is disabled
  2. Sync works with __1.)__ and the default whitelist enabled
  3. Sync works with __2.)__ and `{"url":"firefox.com","options":["authorization"]}` added to the whitelist
  4. Sync doesn't work with __3.)__ and `{"url":"mozilla.com","options":["canvas"]}` added to the whitelist
  4. Sync doesn't work with __2.)__ and `{"url":"mozilla.com","options":["canvas"]}` added to the whitelist

So it seems that __any__ whitelist entry that contains `"url":"mozilla.com"` breaks Firefox Sync. Could it be that adding an URL to the whitelist changes the requests to that url or responses from that URL?

\*) Deactivated means that the checkbox is deactivated, thus the functionality is enabled (e.g. `Disable canvas support` feature deactivated = canvas functionality is enabled)

**Comment by dillbyrne** [comment]
@letorbi Thanks for testing this. I appreciate it. I can't see any reason why it should break it but I can log the urls for firefox sync in the console and see if that can shed some light on the issue. I will do it after I finish #27.

**Comment by dillbyrne** [comment]
@letorbi @rahulgi1  

In my very brief initial test 


To sign in `dom storage` needs to be allowed and the following option was required

`{"url":"accounts.firefox.com","options":["authorization"]}`


when testing sync

All options were selected and the whitelist was not disabled and had the following option  added 

`{"url":"mozilla.com","options":["authorization"]}` 



What worked 

sync from desktop to mobile of bookmarks. I didn't try syncing anything else but if bookmarks work so should passwords and the rest.


What didn't work 

sync from mobile to desktop



I did manage to get mobile to sync once but I was unable to reproduce it. I will look into exact urls  sync uses soon and post an update.

**Comment by letorbi** [comment]
I can confirm that. Bookmark sync from desktop to mobile works, vice versa doesn't.

Maybe the authorization filter gets somehow applied to incoming requests as well?

**Comment by dillbyrne** [comment]
Ok so these are the urls which are used in the sync process

```
 https://accounts.firefox.com
 https://api.accounts.firefox.com
 https://token.services.mozilla.com
 https://sync-183-us-west-2.sync.services.mozilla.com
```

I managed to get everything working (registration and syncing both ways ) by not disabling the authorization header and not disabling dom storage. For this I didn't  have to add anything to the whitelist.


I will try whitelisting the above urls and see if it works with the authorization header disabled

**Comment by dillbyrne** [comment]
When the above urls are whitelisted sync only works from desktop to mobile

**Comment by akwala** [comment]
@dillbyrne @letorbi In your tests, did you already have the Sync credentials set up? They are stored in the Password Mgr as of FX 34. 

With RAS 0.9.5.3 in FX 38, and Sync already set up, I found that leaving Disable Authorization _unchecked_ prevented Sync errors on desktop (didn't test mobile); but setting up Sync only succeeded when I turned off RAS (by selecting Real Profile) -- with RAS on, attempts to set up Sync sowed the not so helpful "Oh dear, somethig is wrong and we have been notified." 

Once sync was set up (as verified by finding the creds in Password Mgr), I turned RAS back on and the synching didn't report errors.

**Comment by dillbyrne** [comment]
@akwala I had the credentials set up in my tests I had to allow dom storage to register for it though.

**Comment by akwala** [comment]
@dillbyrne I allow local DOM storage in RAS, but have configured Self-destructing Cookies to clean it. As I understand it, this should keep the stuff in local storage long enough for ongoing workflow before deleting it.

**Comment by Mosrite** [comment]
@akwala: I can reproduce that!

**Comment by letorbi** [comment]
@akwala I use the same setup.

**Comment by CHEF-KOCH** [comment]
In my case script injection was also the problem (time zone random + protect window Name) rest wasn't checked.

**Comment by Kwaadpepper** [comment]
Disable authorization is breaking as everyone knows but also script injection as said above. Also i was not able to manage the firefox sync account having ```Disable local Dom storage``` as it redirected me on a page saying saying it needs cookies.

**Comment by eNTi** [comment]
this is still an issue. can i do something to help?

**Comment by hovancik** [comment]
Hi guys, so how do iI make it work?  Anyone have it working?

**Comment by dillbyrne** [comment]
Please see #544

**Comment by Sirove** [comment]
> When the above urls are whitelisted sync only works from desktop to mobile

Just want to let you know this is still an issue...
